### ISSUE TITLE
[Bug] Joern tab completion can freeze joern and print "exception caught when loading module class module-info$: java.lang.AssertionError: assertion failed: attempt to parse java.lang.Object from classfile"

### ISSUE THREAD

**Issue by Trolldemorted** [joernio/joern#4625]
**Describe the bug**
Printing `cfg.<tab>` in a linux kernel project makes joern unresponsive and prints the error message `exception caught when loading module class module-info$: java.lang.AssertionError: assertion failed: attempt to parse java.lang.Object from classfile`.

**To Reproduce**
1. Download Ubuntu24's kernel source (linux 6.8.0)
2. Start a joern container with 400GB memory, give 200 of that to joern's xmx
3. Have it parse the linux kernel source code, ignore the occasional `java.lang.OutOfMemoryError: java.lang.OutOfMemoryError: Java heap space`
4. Wait until the interactive shell is back, observe that joern still occupies 200GB memory and 1-10 CPU cores (usage going up and down)
5. Type `cfg.<tab>`  in the shell (probably not the brightest idea I had today)
6. Joern shell is completely unresponsive, joern decides that it now wants 10 cores all the time, let it run some more
7. Press `ctrl+c` a few times
8. See error ("exception caught when loading module class module-info$: java.lang.AssertionError: assertion failed: attempt to parse java.lang.Object from classfile")
9. Wait even more for the shell to be come responsive again

**Expected behavior**
Joern does not become unresponsive through the press of tab, and does not print assertion errors - is this a bad thing? It seems to be scala-specific, all google results I find are related to scala: https://github.com/scala/scala3/issues/20421

**Desktop (please complete the following information):**
 - OS `Ubuntu`
 - Joern/Java Version `ghcr.io/joernio/joern:nightly`

**Comment by DavidBakerEffendi** [comment]
I've experienced this too, does not need that much to reproduce either, usually if one doesn't give an initial character and presses tab this will occur, e.g. `cpg.<TAB>` versus `cpg.c<TAB>`
@mpollmeier  any ideas?

**Comment by mpollmeier** [comment]
Was able to minify it, I'm working on a fix/workaround, there's a few option, bear with me.

**Comment by mpollmeier** [comment]
The root cause is is a scala compiler bug which get's triggered if there's a module-info.class in the classpath - which is a magic JPMS thing generated from the also magic module-info.java that doesn't behave like a regular classfile. 

In joern, we have a module-info.class on the classpath from lombok, which comes in transitively via javasrc2cpg. Some time ago we made the decision to invoke all frontends as external processes (invoking their start scripts), and the more frontends we add, the more important this is. Otherwise we'll end in jar hell with various incompatible versions of many different dependencies, and complex issues with things like OSGI and now JPMS. 

I'll provide a PR (tomorrow) to change the frontends that have been added as classpath dependencies back to process invocations, which will fix this issue.

**Comment by mpollmeier** [comment]
Just took a deeper look at javasrc2cpg and the good thin is that we already invoke it as an external process. At the same time we still have it as a classpath dependency because [depending on the configuration parameters](https://github.com/joernio/joern/blob/706a17ba6f5e53000ff198dff151398ef9b9753d/console/src/main/scala/io/joern/console/cpgcreation/JavaSrcCpgGenerator.scala#L21) we potentially apply some additional passes for type recovery after invoking the frontend: https://github.com/joernio/joern/blob/706a17ba6f5e53000ff198dff151398ef9b9753d/console/src/main/scala/io/joern/console/cpgcreation/JavaSrcCpgGenerator.scala#L27

Question for @johannescoetzee or maybe also @DavidBakerEffendi: wouldn't it make more sense to pass that parameter to the frontend and run those passes as part of the frontend, rather than only in joern? Currently it looks like we have different behaviours between joern's `importCode.java` and running only the frontend itself, e.g. as we do from `sl`. And that sounds like a bad idea.

**Comment by DavidBakerEffendi** [comment]
@mpollmeier Yeah, that initial implementation was a bit hacky to be honest. So the way forward would be to pass it as `--frontend-args` and ingest it again later for post-processing? (so that it can pass between processes)

**Comment by mpollmeier** [comment]
I guess my main question is: why don't we do the post-processing as part of the frontend, so that we get the same results in `sl` and joern?

I assume that there's some legitimate reason for not doing so, but would be good to know either way. Maybe simply because we want the cpg to be as small as possible? 

I'm just thinking: wouldn't it might make more sense to run these additional frontend-specific passes as part of the `importCpg`? Either way, we might need to split out some classes like `JavaTypeRecoveryPassGenerator` into a different project (maybe x2cpg?), so we can remove the dependency from `console->javasrc2cpg`. 

What do you think?

**Comment by DavidBakerEffendi** [comment]
@mpollmeier There are passes SL doesn't want, and they go into post-processing. The reason for this initially was other customers wanted certain features that were implicit in `codescience` or in its data-flow engine. Open-source versions for this functionality were then created and run in post-processing. 

However, if these can simply be disabled by some flag, then that would probably still achieve the same goal.

**Comment by mpollmeier** [comment]
Ah ok, thanks for clarifying, yeah that makes sense. With that in mind I'll see how I'll shuffle things around and add some documentation for the next person wondering the same.

**Comment by mpollmeier** [comment]
https://github.com/joernio/joern/pull/4672

**Comment by mpollmeier** [comment]
this came back with the flatgraph migration, due to another `module-info.class` in a dependency

**Comment by skirge** [comment]
It still happens on my system with JDK 21 (as recommended in README) on Ubuntu 22.04
joern 4.0.78

**Comment by mpollmeier** [comment]
Sorry, somehow I missed that comment. Indeed it wasn't fixed for the distribution, thanks for reporting. This should take care of it though: https://github.com/joernio/joern/pull/5308
